### ISSUE TITLE
(refactor) Too easy to miss enabling a config as in-db and show it on the CLI

### ISSUE THREAD

**Issue by steve-chavez** [PostgREST/postgrest#3101]
## Problem

This occurred on:

- `server-timing-enabled` missing on the CLI https://github.com/PostgREST/postgrest/issues/3100
- `jwt-cache-max-lifetime` missing in-db config on https://github.com/PostgREST/postgrest/pull/3069

This happens because the logic for the CLI and the in-db config is disconnected from the `AppConfig` type

https://github.com/PostgREST/postgrest/blob/0c65bf494d75fbb2b8650eba1385b8d2c5d042d2/src/PostgREST/CLI.hs#L124-L233

https://github.com/PostgREST/postgrest/blob/0c65bf494d75fbb2b8650eba1385b8d2c5d042d2/src/PostgREST/Config/Database.hs#L44-L69

https://github.com/PostgREST/postgrest/blob/0c65bf494d75fbb2b8650eba1385b8d2c5d042d2/src/PostgREST/Config.hs#L70-L115

## Solution

Tie the `AppConfig` or other intermediate type to the CLI an in-db logic.

**Comment by develop7** [comment]
The https://hackage.haskell.org/package/configuration-tools (source at https://github.com/alephcloud/hs-configuration-tools) could be the right tool for the job — it combines CLI options parser, *JSON instances, JSON/YAML config parsing (that's where it falls short for us, though — need to investigate how hard adding custom config formats' support could be), remote configs even, if needed.

**Comment by steve-chavez** [comment]
If we're going to change the file format, I think what would make most sense is using an INI format to be compatible with the [pg service file](https://www.postgresql.org/docs/9.1/libpq-pgservice.html) (also see this [blog post](https://www.cybertec-postgresql.com/en/pg_service-conf-the-forgotten-config-file/)).

That way, we could support multiple databases (https://github.com/PostgREST/postgrest/issues/2798) with a config like:

```ini
# pg_service.ini
[serviceone]
port=5432
user=stevechavez
dbname=stevechavez

[serviceone.postgrest]
jwt-secret = xxxx
db-pool = 10

[servicetwo]
port=5432
...

[servicetwo.postgrest]
jwt-secret = yyy
db-pool = 20
other-configs = ".."
```

This also has the advantage of being able to use psql like:

```
PGSERVICE=serviceone PGSERVICEFILE=pg_service.ini psql

psql (14.10 (Ubuntu 14.10-0ubuntu0.22.04.1))
Type "help" for help.

stevechavez=# 
```

**Comment by steve-chavez** [comment]
> That also means it might be hard to use this file in a different way - i.e. use multiple services from that file at the same time, because that is not what libpq wants to do.

Right, so my intention was to process the `ini` file in PostgREST itself to use multiple dbs. And using the `ini` file we would still have the advantage of leveraging `psql` to test the different connections (`PGSERVICE=anyservice PGSERVICEFILE=our_conf.ini psql`). With YAML we cannot do that.

> The problem for this ini-based file format is, that it will again be impossible to represent anything more complex. YAML can do much more: arrays, objects, anchors, references, ... so much more.

We don't have complex values now and we always need to consider that values have to be compatible with the in-db config. pgbouncer seems to do fine with an [ini config](https://www.pgbouncer.org/config.html) too. So that doesn't seem like a blocker.

**Comment by taimoorzaeem** [comment]
For a short term solution to not miss these, we can calculate the `arity` of `AppConfig` using this [this SO answer](https://stackoverflow.com/a/8369267/16623126). I guess then, we can ensure that length of `dbSettings` list equals `arity AppConfig`, if not then we get a compile-time error.

**Comment by taimoorzaeem** [comment]
> > if not then we get a compile-time error.
> 
> `arity` is only available at run-time, right? How would that be a compile-time check?

Yes, sorry I confused the startup with compile-time. So, at runtime (startup) this would be checked.

**Comment by taimoorzaeem** [comment]
> > I guess then, we can ensure that length of `dbSettings` list equals `arity AppConfig`,
> 
> Not all configuration parameters are available as in-db-config, though. On purpose. So this will never match. What am I missing?

For `in-db-config` that don't have an app-config we'll put `Nothing` in the list. I was thinking we could have a global assertion or at startup like `length maybeDbSettings == arity AppConfig`. Then later use `catMaybes` function to get the actual `dbSettings`. 

This way whenever we introduce a new config the assertion would become false, so we wouldn't miss the db-config. I think slight caveat is that it wouldn't work quite good when replacing or rename a config.

**Comment by taimoorzaeem** [comment]
> We have haskell at our hands, which provides so much benefits with strong typing at compile-time.

True. As far as I know, most compile-time benefits are utilized by using template haskell which we are moving away from, right?

**Comment by taimoorzaeem** [comment]
How about a `Config` type used in combination with kinds and GADTs? So it should look something like:

```haskell
{-# LANGUAGE DataKinds #-}
{-# LANGUAGE KindSignatures #-}
{-# LANGUAGE GADTs #-}

data Config :: Type -> ConfigType -> Type where
  FileConfig         :: Config a 'File
  FileAndInDbConfig  :: Config a 'FileAndInDb

data ConfigType
  = File
  | FileAndInDb

data AppConfig = AppConfig
  { configJwtCacheMaxLifetime    :: Config Int 'FileAndInDb
   ...
  }
```

I think this will force every config to have a `'File` or `'FileAndInDb` value at type level.

**Comment by steve-chavez** [comment]
Not sure if GADTs are strictly needed, but it looks good to me as a first pass, any improvement is welcomed for the Config module.

**Comment by wolfgangwalther** [comment]
How would this enforce that those with type `FileAndInDb` are actually taken from the DB, too?

**Comment by taimoorzaeem** [comment]
> How would this enforce that those with type `FileAndInDb` are actually taken from the DB, too?

We can do this by filtering those fields from the `AppConfig` record and then convert to `[Text]` same way it is in `dbSettingsNames`. How exactly we filter the fields, I am note sure yet, but it is possible.

**Comment by steve-chavez** [comment]
Just remembered how hasql transitioned towards having a config which is a list of settings ([ref](https://github.com/nikita-volkov/hasql-pool/pull/42/changes#diff-4a94efa63532fcf6df9395f5eba174cbde152e00c03de54ac357a686c5384583R24-R30)) instead of having the settings as fields of a type ([AppConfig](https://github.com/PostgREST/postgrest/blob/a525a23c23019fa03a7fd38ab8329d33c1d36248/src/PostgREST/Config.hs#L75)), maybe this can help deduplicating.

**Comment by taimoorzaeem** [comment]
Hmm, not sure if that would help. Looking at [ref](https://github.com/nikita-volkov/hasql-pool/pull/42/changes#diff-aec5c94169fc68724c1518f8fae3a5ad087e4fd9bcfc78d06f2985667a8b27caR22), they have a _default setting_ which is then overridden:

```hs
-- | Compile config from a list of settings.
-- Latter settings override the preceding in cases of conflicts.
settings :: [Setting.Setting] -> Config.Config
settings =
  foldr ($) Config.defaults . fmap Setting.apply
```

Internally, it is still a record type:

```hs
-- | Reasonable defaults, which can be built upon.
defaults :: Config
defaults =
  Config
    { size = 3,
      acquisitionTimeout = 10,
      agingTimeout = 60 * 60 * 24,
      idlenessTimeout = 60 * 10,
      connectionSettingsProvider = pure "postgresql://postgres:postgres@localhost:5432/postgres",
      observationHandler = const (pure ())
    }
```

**Comment by steve-chavez** [comment]
Another problem with the config type, we have to duplicate default values in the spec tests :

https://github.com/PostgREST/postgrest/blob/a525a23c23019fa03a7fd38ab8329d33c1d36248/src/PostgREST/Config.hs#L497

https://github.com/PostgREST/postgrest/blob/a525a23c23019fa03a7fd38ab8329d33c1d36248/test/spec/SpecHelper.hs#L130
